[Tests] Create a GuiTester to dramatically increase code coverage

NicolasDorier opened the pull request:
My refactoring on removing global state has been done to be able to easily create unit tests which setup a full wasabi wallet environment.

Example:
```csharp
		[Fact]
		public async Task CanLoadCoinsAsync()
		{
			using (var tester = GuiTester.Create())
			{

 				var client1 = tester.CreateGuiClient();
				var client2 = tester.CreateGuiClient();
				await tester.StartAllAsync();
				await Task.Delay(10000);
			}
		}
```

Via client1 and 2, you can access the mainviewmodel and global GUI.

This is far from being ready to merge, my goal is to reach the state where some basic operations are tested like: 
* If client1 and client2 receives coins, does it appears correctly in history and coin list?
* Can client1 properly send money to client2 via the view models?

The two clients have their own independant datadir, with their own full nodes.

I will then do tests on simple mixing and test how the viewmodels behave.

With such framework, it will become very easy to create 10 clients mixing together while reaching very high code coverage.

I need to use NBitcoin.Testframework, because of the reliance of Wasabi one on singletons which would prevent me to do such features easily.

NicolasDorier commented on WalletWasabi.Gui/WalletWasabi.Gui.csproj:
```diff
@@ -67,5 +67,14 @@
     <RuntimeHostConfigurationOption Include="System.Globalization.Invariant" Value="true" />
   </ItemGroup>
 
+  <ItemGroup>
```
unsure where does it come from. Will remove later.

nopara73 replied:
Yes please, it's some Avalonia issue, we keep constantly removing this accidental csproj trashing.

danwalmsley replied:
I think you mean visual studio issue ;)

nopara73 commented on WalletWasabi.Tests/WalletWasabi.Tests.csproj:
```diff
@@ -16,6 +16,7 @@
     <PackageReference Include="AltCover" Version="5.3.675" />
     <PackageReference Include="Microsoft.AspNetCore.App" />
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.2.0" />
+    <PackageReference Include="NBitcoin.TestFramework" Version="1.7.6" />
```
Please don't add packages to the project.

NicolasDorier replied:
I tried to use Wasabi's `CoreNode` from Wasabi.Tests instead but its design make it impossible to use properly in my case. Mainly due to the use of the Test Global and the fact there is no way to separate the node creation (`Create`, which should not start the node) and the start of the node.

Would you agree for either: 
* Removing CoreNode from Wasabi.Tests and use NBitcoin.TestFramework instead everywhere.
* or let me refactor the Wasabi one so that we don't end up with singletons dependencies.

nopara73 commented on WalletWasabi.Gui/Program.cs:
```diff
@@ -5,11 +5,13 @@
 using WalletWasabi.Gui.CommandLine;
 using WalletWasabi.Gui.ViewModels;
 using WalletWasabi.Logging;
+[assembly:InternalsVisibleTo("WalletWasabi.Tests")]
```
Please remove this and make the internal public, we don't use internals, only by accident.

NicolasDorier replied:
I think this is wrong. Making tests internal is a practice used widely by Microsoft themselves. (https://github.com/aspnet/AspNetCore/search?q=InternalsVisibleTo&unscoped_q=InternalsVisibleTo)

nopara73 commented:
@danwalmsley Is this the same pattern you wanted to test the GUI back then? Maybe if we decide to do that, we should use the correct pattern.

NicolasDorier commented:
Still early but the current progress is:
* Coordinator running with his node,
* Each GUI client runs with their node and connect to the coordinator (at Bitcoin P2P and Wasabi protocol level) successfully.

danwalmsley commented:
> @danwalmsley Is this the same pattern you wanted to test the GUI back then? Maybe if we decide to do that, we should use the correct pattern.

No, although im sure this probably tests other stuff.

I would recommend as is standard for MVVM projects to add unit tests around the ViewModels to test critical UI logic.

 This may need some kind of stubbing out of backend layer to interfaces or some simulated global interface.

For example you might test that a ViewModel that deals with email addresses:

1) correctly validates an email address.
2) displays error if validation fails
3) disables Submit form button command unless the email address is valid.

That kind of thing.

NicolasDorier commented:
@danwalmsley that's the goal, this this PR the `client1` has the `MainViewModel` accessible.
I think it won't be too difficult after I get some other PR merged.

lontivero commented:
@NicolasDorier is this waiting for something?

NicolasDorier commented:
@lontivero yes, me rebasing it. Will do.

NicolasDorier commented:
I rebased, ready for review, the fact that we removed singleton now mean we can effectively create several gui wallet and backend at the same time.

There is a `Logs.txt` file created in the test folder helping for debugging.

This first test is pretty simple: Make sure that the Gui wallet can initialize properly.

lontivero commented on WalletWasabi.Tests/GuiTester.cs:
```diff
@@ -0,0 +1,252 @@
+
+		private static void TaskScheduler_UnobservedTaskException(object sender, UnobservedTaskExceptionEventArgs e)
+		{
+			e.Exception.ToString();
```
What is the idea here? This line does nothing.

lontivero commented on WalletWasabi.Tests/GuiTester.cs:
```diff
@@ -0,0 +1,252 @@
+		}
+
+		private static void CurrentDomain_UnhandledException(object sender, UnhandledExceptionEventArgs e)
+		{
```
Shouldn't we write the exception to the console/log?

NicolasDorier commented:
Fix the log of unhandled exceptions.

lontivero commented:
I've reviewed it and tested it (I also have ran the test in debug mode just to understand what it does). It works okay, test pass. The only problem that I see it that it uses the NBitcoin's `CoreNode` instead of the Wasabi's one. I really don't remember why we started using our own version (there was a good reason by sure).

So, something has to be done about this. IMO, it would be great to be able to use the NBitcoin `CoreNode` and drop the one that we have but that's something we need to discuss first. Ping @nopara73
